shelley/design-spec/delegation_design_spec.tex

```diff
@@ -567,20 +574,63 @@ \subsection{Overview of Delegation}
 \subsection{Address Structure}
 \label{address-structure}
 
+To separate the concerns of moving funds and exercising rights to participate in
+the protocol, we discern between different types of addresses:
+\begin{description}
+\item[Value Addresses] are used to transfer value. Whenever a transaction spends
+  a UTxO entry, it has to contain a witness for the value address of that UTxO
+  entry.
+
+  In its simplest form, a value address is associated with a payment key pair
+  \((skp, vkp)\). The address then consists of a cryptographic hash
+  \(\hash{vkp}\) of the verification key \(vkp\). A witness for such an address
+  contains the verification key \(vkp\) and a signature of the signing key
+  \(skp\). We call those addresses \emph{single key addresses} or simply
+  \emph{key addresses}, and the corresponding witnesses \emph{single key
+    witnesses} or \emph{key witnesses}.
+
+  We also support value addresses that need signatures from multiple parties.
+  Those also contain a hash, not of a verification key, but of a validator
+  script. The script can be written either in Plutus, or in a mini-language for
```

Comment:
If the aim is for things to be uniform between the mini-language and Plutus, what happens when we go to full UTXO/extended UTXO? Will the mini-language script also get passed some kind of redeemer? Or are multi-sig addresses going to be distinct from "full script" addresses, and we'll just have a different validation function for them?

Reply:
The latter.

What I have in my head for integrating Plutus is that we'll use one address type for all kinds of scripts (the hash of a validator script). What's passed to the validator script depends on the language. Scripts in a special-purpose multi-sig language will only expect witnesses for some of the keys, while Plutus scripts will also get a redeemer script and data from the extended UTxO.

Does that make sense?

Reply:
That suggests that we won't be able to write the ledger rules for script addresses without having an enumeration of all the languages, since the interface to each might be different. Maybe not a problem.

Reply:
Yes, we expect that the set of info needed as input for different languages will be different. I think that's ok since the witness tells us which language is being used (which we can then verify against the hash in the address) and then depending on the language we can pass in the right things to the interpreter (things from the witness like the redeemer, info about the tx, other environment info).

So yes, the inputs to the script interpreter is expected to be non-uniform, with cases for each language, with the language tag given by the witness not the script address.